### PR TITLE
Sindhu/enable multiple batches for cpp examples 

### PR DESCRIPTION
**Pull request by sindhu-nervana**

### CODE REVIEW

File: examples/cpp/infer_multiple_networks/main.cc [hunk 1]
```diff
@@ -89,18 +89,21 @@ void PrintVersion() {
 }
 
 int main(int argc, char** argv) {
-  string image = "grace_hopper.jpg";
-  string graph = "inception_v3_2016_08_28_frozen.pb";
+  // parameters below need to modified as per model
+  string image = "3.png";
+  // Size of the vector images will be batch size
+  vector<string> images(12, image);
```
Comment (shresthamalik) [inline examples/cpp/infer_multiple_networks/main.cc]: Please make batch size a variable, that can be set in the main

File: examples/cpp/infer_single/infer_single.cpp [hunk 2]
```diff
@@ -41,11 +42,12 @@ extern tf::Status LoadGraph(const string& graph_file_name,
                             std::unique_ptr<tf::Session>* session,
                             const tf::SessionOptions& options);
 
-extern tf::Status ReadTensorFromImageFile(const string& file_name,
+extern tf::Status ReadTensorFromImageFile(const std::vector<string>& file_name,
                                           const int input_height,
                                           const int input_width,
                                           const float input_mean,
                                           const float input_std, bool use_NCHW,
+                                          const int wanted_channels,
```
Comment (shresthamalik) [inline examples/cpp/infer_single/infer_single.cpp]: Rename wanted_channels to input_channels

File: examples/cpp/infer_single/infer_single.cpp [hunk 3]
```diff
@@ -133,7 +135,10 @@ std::unique_ptr<tf::Session> CreateSession(const string& graph_filename) {
 }
 
 int main(int argc, char** argv) {
+  // parameters below need to modified as per model
   string image = "image_00000.png";
+  // Size of the vector images will be batch size
+  std::vector<string> images(12, image);
```
Comment (shresthamalik) [inline examples/cpp/infer_single/infer_single.cpp]: same thing. Batch size can be made a variable

File: examples/cpp/infer_multiple_networks/inference_engine.cc [hunk 4]
```diff
@@ -40,28 +40,29 @@ extern tf::Status LoadGraph(const string& graph_file_name,
...
 namespace infer_multiple_networks {
 
 InferenceEngine::InferenceEngine(const string& name, const string& backend)
     : m_name(name) {}
-Status InferenceEngine::Load(const string& network, const string& image_file,
-                             int input_width, int input_height,
-                             float input_mean, float input_std,
-                             const string& input_layer,
+Status InferenceEngine::Load(const string& network,
+                             std::vector<string>& image_files, int input_width,
```
Comment (sayantan-nervana) [inline examples/cpp/infer_multiple_networks/inference_engine.cc]:
```suggestion
                             const std::vector<string>& image_files, int input_width,
```

File: examples/cpp/infer_multiple_networks/inference_engine.cc [hunk 5]
```diff
@@ -40,28 +40,29 @@ extern tf::Status LoadGraph(const string& graph_file_name,
...
+                             int input_height, float input_mean,
+                             float input_std, const string& input_layer,
                              const string& output_layer, bool use_NCHW,
-                             bool preload_images) {
+                             bool preload_images, const int input_channels) {
```
Comment (sayantan-nervana) [inline examples/cpp/infer_multiple_networks/inference_engine.cc]: Minor
in this function signature some things are marked const others not. eg "input_channels" is marked const.
maybe make everything that can be: `const T&`

File: examples/cpp/tf_label_image_utils.cc [hunk 6]
```diff
@@ -144,69 +144,80 @@ static Status ReadEntireFile(tensorflow::Env* env, const string& filename,
 // Given an image file name, read in the data, try to decode it as an image,
 // resize it to the requested size, and then scale the values as desired.
 //-----------------------------------------------------------------------------
-Status ReadTensorFromImageFile(const string& file_name, const int input_height,
-                               const int input_width, const float input_mean,
-                               const float input_std, bool use_NCHW,
+Status ReadTensorFromImageFile(const std::vector<string>& file_names,
+                               const int input_height, const int input_width,
+                               const float input_mean, const float input_std,
+                               bool use_NCHW, const int input_channels,
```
Comment (sayantan-nervana) [inline examples/cpp/tf_label_image_utils.cc]: Minor
similarly `const int` might as well be `const int&`

File: examples/cpp/tf_label_image_utils.cc [hunk 7]
```diff
@@ -144,69 +144,80 @@ static Status ReadEntireFile(tensorflow::Env* env, const string& filename,
...
+    } else {
+      // Assume if it's neither a PNG nor a GIF then it must be a JPEG.
+      image_reader = DecodeJpeg(root.WithOpName("jpeg_reader"), file_reader,
+                                DecodeJpeg::Channels(input_channels));
+    }
+    // Now cast the image data to float so we can do normal math on it.
+    auto float_caster = Cast(root.WithOpName("float_caster"), image_reader,
+                             tensorflow::DT_FLOAT);
+    // The convention for image ops in TensorFlow is that all images are
+    // eƒixpected
```
Comment (sayantan-nervana) [inline examples/cpp/tf_label_image_utils.cc]:
```suggestion
    // expected
```

File: examples/cpp/infer_multiple_networks/inference_engine.cc [hunk 8]
```diff
@@ -40,28 +40,30 @@ extern tf::Status LoadGraph(const string& graph_file_name,
...
-                             int input_width, int input_height,
-                             float input_mean, float input_std,
+Status InferenceEngine::Load(const string& network,
+                             const std::vector<string>& image_files,
+                             const int& input_width, const int& input_height,
+                             const float& input_mean, const float& input_std,
                              const string& input_layer,
                              const string& output_layer, bool use_NCHW,
-                             bool preload_images) {
+                             bool preload_images, const int& input_channels) {
```
Comment (avijit-nervana) [inline examples/cpp/infer_multiple_networks/inference_engine.cc]: Why do you have const int&? just an `int` would do - no?

File: examples/cpp/infer_multiple_networks/inference_engine.h [hunk 9]
```diff
@@ -44,10 +45,11 @@ class InferenceEngine {
...
   ~InferenceEngine();
 
-  Status Load(const string& network, const string& image_file, int input_width,
-              int input_height, float input_mean, float input_std,
+  Status Load(const string& network, const std::vector<string>& image_files,
+              const int& input_width, const int& input_height,
+              const float& input_mean, const float& input_std,
               const string& input_layer, const string& output_layer,
-              bool use_NCHW, bool preload_images);
+              bool use_NCHW, bool preload_images, const int& input_channels);
```
Comment (avijit-nervana) [inline examples/cpp/infer_multiple_networks/inference_engine.h]: Why do you have const int&? just an int would do - no?

File: examples/cpp/infer_multiple_networks/main.cc [hunk 10]
```diff
@@ -89,18 +89,22 @@ void PrintVersion() {
...
-  string image = "grace_hopper.jpg";
-  string graph = "inception_v3_2016_08_28_frozen.pb";
+  // parameters below need to modified as per model
+  string image = "3.png";
+  int batch_size = 12;
+  // Vector size is same as the batch size, populating with single image
+  vector<string> images(batch_size, image);
+  string graph = "mnist_inference_quantized_trained_12212018.pb";
   string labels = "";
-  int input_width = 299;
```
Comment (avijit-nervana) [inline examples/cpp/infer_multiple_networks/main.cc]: Please restore the previous example as the mnist_inference_quantized_trained_12212018.pb is not publicly available. Just use the BS=1 as the default.

File: examples/cpp/infer_multiple_networks/inference_engine.cc (hunk 8 above)
Comment (sindhu-nervana) [reply examples/cpp/infer_multiple_networks/inference_engine.cc]: It would. Few were const T& and few weren't here so made all of them same, will modify.

**Pull request merged**